File: 2017/6/main.py
with open('input') as reader:
    input = reader.read()

allStates = []

def processInput():
    temp = input.split("\t")
    temp = list(map(int, temp))
    return temp

def getMax(num_list):
    temp = sorted(num_list)
    size = len(temp)
    maxElem = temp[size-1]
    position = num_list.index(maxElem)
    return maxElem, position

def isPresent(current):
    flag = True
    
    try:
        allStates.index(current)
    except ValueError:
        flag = False
    
    return flag

def redistribute(state, largest, position):
    n = 0
    pointer = position + 1
    state[position] = 0
    size = len(state)

    while n < largest:
        state[pointer % size] += 1
        pointer += 1
        n += 1

    return state

def fun(inputs, count):
    while True:
        if isPresent(inputs):
            print(count)
            print(count - allStates.index(inputs))
            break
        else:
            input_copy = inputs.copy()
            large, pos = getMax(input_copy)
            count+=1
            red = redistribute(input_copy, large, pos)
            allStates.append(inputs)
            inputs = red

if __name__ == "__main__":
    int_input = processInput()
    # int_input = [0,2,7,0]
    fun(int_input, 0)
    

File: 2017/6/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"reflect"
	"strconv"
	"strings"
)

type stateVar struct {
	stateArr []int
}

func isError(e error) {
	if e != nil {
		panic(e)
	}
}

func wasDone(state [][]int, current []int) bool {
	flag := false

	for i := range state {
		if reflect.DeepEqual(state[i], current) {
			flag = true
			break
		}
	}

	return flag
}

func getMax(arr []int) (int, int) {
	max := arr[0]
	index := 0

	for i, num := range arr {
		if num > max {
			max = num
			index = i
		}
	}

	return max, index
}

func redistribute(arr []int, amount int, index int) []int {
	temp := arr
	n := 0

	pointer := index + 1

	temp[index] = 0

	size := len(arr)

	for n < amount {
		temp[pointer%size]++
		pointer++
		n++
	}

	return temp
}

func main() {
	file, err := os.Open("input")
	isError(err)
	defer file.Close()

	scanner := bufio.NewScanner(file)

	state := []int{}

	init := []string{}

	for scanner.Scan() {
		init = strings.Split(scanner.Text(), "\t")
	}

	allStates := []stateVar{}

	for i := range init {
		number, err := strconv.Atoi(init[i])
		isError(err)
		state = append(state, number)
	}

	count := 0

	for { // THIS PROCESS CAN BE EASILY CONVERTED TO A RECURSIVE PROCESS
		currentLargest, position := getMax(state)

		count++

		fmt.Println(state)

		allStates = append(allStates, stateVar{stateArr: redistribute(state, currentLargest, position)})

		fmt.Println("State : ", allStates)

		if count > 5 {
			break
		}

		// if wasDone(allStates, temp) {
		// 	break
		// } else {q
		// 	allStates = append(allStates, temp)
		// 	state = temp
		// 	fmt.Println(state, allStates)
		// }
	}

	fmt.Println(count-1, allStates)
}
